### PR TITLE
 Use IOT sockets instead of lwip directly #6 

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**

### CODE REVIEW

**Comment by ATmobica** [comment]
[Feature] I think we can implement our signal socket as an alternative `WakeEvent()` class to be closer to the Matter approach. But not in this iteration.

File: src/inet/IPAddress-StringFuncts.cpp [hunk 1]
```diff
@@ -81,6 +85,21 @@ char * IPAddress::ToString(char * buf, uint32_t bufSize) const
 #elif CHIP_SYSTEM_CONFIG_USE_OPEN_THREAD_ENDPOINT
     otIp6Address addr = ToIPv6();
     otIp6AddressToString(&addr, buf, static_cast<uint16_t>(bufSize));
+#elif CHIP_SYSTEM_CONFIG_USE_IOT_SOCKET
+#if INET_CONFIG_ENABLE_IPV4
+    if (IsIPv4())
+    {
+        ip4_addr_t ip4_addr;
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/IPAddress-StringFuncts.cpp:92]: I would have to then provide my own to string function

File: src/inet/IPAddress.cpp [hunk 2]
```diff
@@ -97,6 +97,7 @@ IPAddress::IPAddress(const ip_addr_t & addr)
 
 #endif // INET_CONFIG_ENABLE_IPV4 || LWIP_IPV4
 
+#if !CHIP_SYSTEM_CONFIG_USE_IOT_SOCKET
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/IPAddress.cpp:100]: they were needed at the beginning, you're right I can probably remove them now

File: src/inet/TCPEndPointImplIoTSocket.cpp [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+CHIP_ERROR TCPEndPointImplIoTSocket::BindImpl(IPAddressType addrType, const IPAddress & addr, uint16_t port, bool reuseAddr)
+{
+    CHIP_ERROR res = GetSocket(addrType);
+
+    if (res == CHIP_NO_ERROR && reuseAddr)
+    {
+        uint32_t n = 1;
+        iotSocketSetOpt(mSocket, IOT_SOCKET_SO_REUSEADDR, &n, sizeof(n));
+
+#ifdef SO_REUSEPORT
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/TCPEndPointImplIoTSocket.cpp:53]: double call is a mistake, thanks, good question about the ifdefs, I guess it means those options are not needed

File: config/openiotsdk/chip-gn/args.gni [hunk 4]
```diff
@@ -27,6 +27,7 @@ chip_config_network_layer_ble = false
 chip_system_config_use_lwip = true
 lwip_platform = "external"
 chip_system_config_use_sockets = false
+chip_system_config_use_iot_socket = true
```
[the remark this answers is not in the document]
Comment (pan-) [reply config/openiotsdk/chip-gn/args.gni:30]: I think it should be configurable.

File: src/system/SystemLayerImplOpenIoTSDK.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            current++;
+        }
+        return false;
+    };
+
+    static constexpr int kSocketMax = (INET_CONFIG_ENABLE_TCP_ENDPOINT ? INET_CONFIG_NUM_TCP_ENDPOINTS : 0) +
+        (INET_CONFIG_ENABLE_UDP_ENDPOINT ? INET_CONFIG_NUM_UDP_ENDPOINTS : 0);
+
+    chip::Inet::EndPointStateIoTSocket * mSelectEndpoints[kSocketMax];
+    SelectMask mMaskMemory        = nullptr;
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/system/SystemLayerImplOpenIoTSDK.h:110]: no, each layer can have different sockets waiting (to be honest the whole idea of have multiple system layers is wrong)

File: src/inet/UDPEndPointImplIoTSocket.cpp [hunk 6]
```diff
@@ -0,0 +1,751 @@
...
+            {
+#if ENABLE_PKTINFO
+#if INET_CONFIG_ENABLE_IPV4
+                if (controlHdr->cmsg_level == IOT_SOCKET_LEVEL_IPPROTO_IP && controlHdr->cmsg_type == IOT_SOCKET_IP_PKTINFO)
+                {
+                    iot_in_pktinfo * inPktInfo = (iot_in_pktinfo *) IOT_CMSG_DATA(controlHdr);
+                    InterfaceId iface;
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+                    char name[IF_NAMESIZE];
+                    if (lwip_if_indextoname(inPktInfo->ipi_ifindex, name))
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/UDPEndPointImplIoTSocket.cpp:556]: No, we're just translating index to pointer - our implementation doesn't store the intex, it stores the pointer to the netif struct

File: src/inet/UDPEndPointImplIoTSocket.cpp [hunk 7]
```diff
@@ -0,0 +1,751 @@
...
+    // address, construct an IOT_SOCKET_IP_PKTINFO/IOT_SOCKET_IPV6_PKTINFO "control message" to that effect
+    // add add it to the message header.  If the local OS doesn't support IOT_SOCKET_IP_PKTINFO/IOT_SOCKET_IPV6_PKTINFO
+    // fail with an error.
+
+    // iotSocketSendMsg doesn't support pktinfo, we can't specify the address, fail if we try to
+    if (aPktInfo->SrcAddress.Type() != IPAddressType::kAny)
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+    else if (intf.IsPresent())
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/UDPEndPointImplIoTSocket.cpp:305]: this is to catch errors in the future, when we have more than one interface

[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/UDPEndPointImplIoTSocket.cpp:305]: other implementations don't have the same limitations as us

File: src/inet/UDPEndPointImplIoTSocket.cpp [hunk 8]
```diff
@@ -0,0 +1,751 @@
...
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+#include "lwip/if_api.h"
+#include "lwip/ip4.h"
+#include "lwip/ip6.h"
+#endif // CHIP_SYSTEM_CONFIG_USE_LWIP
+
+#define ENABLE_PKTINFO 1
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply src/inet/UDPEndPointImplIoTSocket.cpp:33]: oh, this was only useful during development, removing

File: src/inet/IPAddress-StringFuncts.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ATmobica) [reply src/inet/IPAddress-StringFuncts.cpp:92]: ok, understand, there is no `xxx_ntoa_r` function implementation in the IoT socket. I just want to reduce the direct use of LwiP as much as possible.

**Pull request closed without merging**